fairlearn/reductions/_exponentiated_gradient/_lagrangian.py

```diff
@@ -56,39 +56,39 @@ def __init__(self, X, sensitive_features, y, estimator, constraints, eps, B, opt
         self.last_linprog_n_hs = 0
         self.last_linprog_result = None
 
-    def _eval_from_error_gamma(self, error, gamma, lambda_vec):
+    def _eval(self, h, lambda_vec):
         """Return the value of the Lagrangian.
 
-        :return: tuple `(L, L_high)` where `L` is the value of the Lagrangian, and
-            `L_high` is the value of the Lagrangian under the best response of the lambda player
-        :rtype: tuple of two floats
+        :param h: TODO: should this be Q instead of h? It seems to indicate how much weight each h
+            gets in the randomized classifier, summing up to 1
+        :type h: pandas.Series
+        :param lambda_vec: lambda vector
+        :type h: pandas.Series
+
+        :return: tuple `(L, L_high, gamma, error)` where `L` is the value of the Lagrangian,
+            `L_high` is the value of the Lagrangian under the best response of the lambda player,
+            `gamma` is the vector of constraint violations, and `error` is the empirical error
         """
-        lambda_projected = self.constraints.project_lambda(lambda_vec)
+        error = self.errors[h.index].dot(h)
```

Comment:
Let's revert this piece of code to what it was previously, meaning:
```python
        if callable(h):
            error = self.obj.gamma(h)[0]
            gamma = self.constraints.gamma(h)
        else:
            error = self.errors[h.index].dot(h)
            gamma = self.gammas[h.index].dot(h)
```
I think that the evaluation of Lagrangian should work for both h and Q.

Reply:
And you can of course update the doc string at the top appropriately--that h can be either a callable or a Series representing Q.

Reply:
Are you aware that it's never called with callable(h)? It's a dead code path.

Reply:
I know it's a dead code right now, but let's keep it--we could even use it in tests with a callable rather than Q.

Reply:
If we want to keep it around, it would be best that there's a test which touches it somehow. Otherwise, there's a tendency for the code to 'drift' and you eventually find out that the code which you've been keeping around 'in case it's useful' doesn't actually work any more.

Reply:
added back + tests for the callable(h) case